x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -184,59 +185,98 @@ export class ExecutionHandler<
-              alertId: ruleId,
-              alertType: this.ruleType.id,
-              actionTypeId,
-              alertName: this.rule.name,
+        if (alert) {
+          const actionToRun = {
+            ...action,
+            params: injectActionParams({
+              ruleId,
               spaceId,
-              tags: this.rule.tags,
-              alertInstanceId: alert.getId(),
-              alertActionGroup: actionGroup,
-              alertActionGroupName: this.ruleTypeActionGroups!.get(actionGroup)!,
-              context: alert.getContext(),
-              actionId: action.id,
-              state: alert.getScheduledActionOptions()?.state || {},
-              kibanaBaseUrl: this.taskRunnerContext.kibanaBaseUrl,
-              alertParams: this.rule.params,
-              actionParams: action.params,
-              ruleUrl: this.buildRuleUrl(spaceId),
+              actionTypeId,
+              actionParams: transformActionParams({
+                actionsPlugin,
+                alertId: ruleId,
+                alertType: this.ruleType.id,
+                actionTypeId,
+                alertName: this.rule.name,
+                spaceId,
+                tags: this.rule.tags,
+                alertInstanceId: alert.getId(),
+                alertActionGroup: actionGroup,
+                alertActionGroupName: this.ruleTypeActionGroups!.get(actionGroup)!,
+                context: alert.getContext(),
+                actionId: action.id,
+                state: alert.getScheduledActionOptions()?.state || {},
+                kibanaBaseUrl: this.taskRunnerContext.kibanaBaseUrl,
+                alertParams: this.rule.params,
+                actionParams: action.params,
+                ruleUrl: this.buildRuleUrl(spaceId),
+              }),
             }),
-          }),
-        };
-
-        await this.actionRunOrAddToBulk({
-          enqueueOptions: this.getEnqueueOptions(actionToRun),
-          bulkActions,
-        });
-
-        logActions.push({
-          id: action.id,
-          typeId: action.actionTypeId,
-          alertId: alert.getId(),
-          alertGroup: action.group,
-        });
-
-        if (this.isRecoveredAlert(actionGroup)) {
-          alert.scheduleActions(action.group as ActionGroupIds);
-        } else {
-          if (action.frequency?.throttle) {
-            alert.updateLastScheduledActions(
-              action.group as ActionGroupIds,
-              this.generateActionHash(action)
-            );
+          };
+
+          await this.actionRunOrAddToBulk({
+            enqueueOptions: this.getEnqueueOptions(actionToRun),
+            bulkActions,
+          });
+
+          logActions.push({
+            id: action.id,
+            typeId: action.actionTypeId,
+            alertId: alert.getId(),
+            alertGroup: action.group,
+          });
+
+          if (this.isRecoveredAlert(actionGroup)) {
+            alert.scheduleActions(action.group as ActionGroupIds);
```

Comment:
Do we need this? we keep this here because it has always been after a recovered alert execution, but nothing happens when i remove it...

Reply:
From looking at the code, it would "trigger" a recovered event for a given alert. But I'm guessing this is already happening prior to this code getting called..

Reply:
I think this is actually causing an error when a rule has multiple recovery actions attached. Since we're iterating over actions first, for a recovered alert, the `scheduleActions` gets called multiple times, once for each action and this throws an error in the task runner.

Reply:
It comes from:
https://github.com/elastic/kibana/blob/3b44d76b620e5f5c7d07b8070b5609692745f159/x-pack/plugins/alerting/server/task_runner/schedule_actions_for_alerts.ts#L81

I am not sure if it does something, is removing that enough? or should we find a way to schedule a recovered action after all the recovered alerts are executed...